### PR TITLE
Improve test coverage of sink values for trusted types [part1].

### PR DESCRIPTION
**Pull request by ziransun**
#### 885b108aec57dca7af8f4b473fdbb274b79d6f54
<pre>
Improve test coverage of sink values for trusted types [part1].
<a href="https://bugs.webkit.org/show_bug.cgi?id=273763">https://bugs.webkit.org/show_bug.cgi?id=273763</a>

Reviewed by Manuel Rego Casasnovas.

Check the sink value for defaul policy in the tests.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-Document-execCommand.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-DOMWindowTimers-setTimeout-setInterval.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/modify-attributes-in-callback.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/require-trusted-types-for.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-createHTMLDocument.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-duplicate-names-without-enforcement.html:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/worker-constructor.https.html:

Canonical link: <a href="https://commits.webkit.org/279045@main">https://commits.webkit.org/279045@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/d2b359121b5baf96cc53645a350c78d3c1f0e18c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51764 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31076 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4128 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55030 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2456 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37447 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2152 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42116 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/1514 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53863 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28705 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44652 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23246 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25995 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1912 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/650 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47954 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2004 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56623 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26885 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1881 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49521 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28122 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44740 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48760 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11432 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29019 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27859 "Built successfully") | | |

### CODE REVIEW

File: LayoutTests/imported/w3c/web-platform-tests/trusted-types/modify-attributes-in-callback.html [hunk 1]
```diff
@@ -13,11 +13,14 @@
 
   let target = "data-x";
   trustedTypes.createPolicy("default", {
-    createHTML: (s) => {
+    createHTML: (s, _, sink) => {
+    // TODO: sink value?
+    assert_equals(sink, '');
```
Comment (lukewarlow) [inline LayoutTests/imported/w3c/web-platform-tests/trusted-types/modify-attributes-in-callback.html]: I think the sink value here per spec should be "HTMLIFrameElement srcdoc". This needs updating in the WebKit implementation AND the Chromium implementation.

File: LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-createHTMLDocument.html [hunk 2]
```diff
@@ -56,7 +56,11 @@
 promise_test(t => {
   return new Promise(resolve => {
     trustedTypes.createPolicy("default",
-                              { createHTML: s => s + " [default]" });
+                              { createHTML: (s, _, sink) => {
+                                assert_equals(sink, 'Element innerHTML');
+                                return s + " [default]";
+                              }});
```
Comment (lukewarlow) [inline LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-createHTMLDocument.html]: nit: Indentation looks off might be worth fixing it as you're updating the code anwyay

File: LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-svg-script-set-href.html [hunk 3]
```diff
@@ -71,7 +71,20 @@
     // but now expect all of them to pass.
     promise_test(t => {
       trustedTypes.createPolicy("default", {
-        createScript: x => x, createHTML: x => x, createScriptURL: x => x });
+        createScript: (x, _, sink) => {
+          //TODO: check sink value.
+          assert_equals(sink, 'HTMLScriptElement src');
```
Comment (lukewarlow) [inline LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-svg-script-set-href.html]: My feeling is that this should be `"SVGScriptElement text"`. But I think it depends what actually triggers this. Maybe run the test in Chromium and see what it actually gets

File: LayoutTests/platform/mac-ventura/TestExpectations [hunk 4]
```diff
@@ -1,3 +1,4 @@
 # Failing after OS migration rdar://112624778 (Migrate macOS Sonoma test expectations to OpenSource, add expectation files to Down-Levels (259373))
 
+webkit.org/b/266630 imported/w3c/web-platform-tests/trusted-types/trusted-types-createHTMLDocument.html [ Pass Failure ]
```
Comment (lukewarlow) [inline LayoutTests/platform/mac-ventura/TestExpectations]: Move this to the top-level TestExpectations file and link it to https://bugs.webkit.org/show_bug.cgi?id=272196

File: LayoutTests/imported/w3c/web-platform-tests/trusted-types/modify-attributes-in-callback.html (hunk 1 above)
Comment (ziransun) [reply LayoutTests/imported/w3c/web-platform-tests/trusted-types/modify-attributes-in-callback.html]: Thanks. Will check both implementations.

Comment (ziransun) [reply LayoutTests/imported/w3c/web-platform-tests/trusted-types/modify-attributes-in-callback.html:18]: I think we might need to add `sink` paramenter in `DOMString? [getAttributeType](https://w3c.github.io/trusted-types/dist/spec/#dom-trustedtypepolicyfactory-getattributetype)(`

Comment (ziransun) [reply LayoutTests/imported/w3c/web-platform-tests/trusted-types/modify-attributes-in-callback.html]: You are right. I changed the implementation at https://github.com/WebKit/WebKit/pull/26519 so now we have the sink value passed over as "HTMLIFrameElement srcdoc".

File: LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-svg-script-set-href.html (hunk 3 above)
Comment (ziransun) [reply LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-svg-script-set-href.html]: Chromium gives "Element setAttribute".

Comment (lukewarlow) [reply LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-svg-script-set-href.html]: In that case that's Chrome not matching latest spec and so I think it should be `SVGScriptElement text` feel free to leave this one for now though until we know what it should be. (once my script protection PR is in we should get our answer)

Comment (ziransun) [reply LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-svg-script-set-href.html]: As discussed, I'll move this test to a different patch together with other relevant scriptElement tests.

File: LayoutTests/imported/w3c/web-platform-tests/trusted-types/worker-constructor.https.html [hunk 5]
```diff
@@ -64,7 +64,16 @@
...
 promise_test(t => {
   trustedTypes.createPolicy("default", {
-      createScriptURL: s => s.replace("potato", "https") });
+    createScriptURL: (s, _, sink) => {
+      if (s === "Worker") {
+        assert_equals(sink, 'Worker create');
+      } else if (s === "SharedWorker") {
+        assert_equals(sink, 'SharedWorker create');
+      } else if (s == "service_worker") {
+        assert_equals(sink, 'ServiceWorkerContainer addRegistration');
```
Comment (lukewarlow) [inline LayoutTests/imported/w3c/web-platform-tests/trusted-types/worker-constructor.https.html]: `ServiceWorkerContainer register` I think it should be this? Per https://w3c.github.io/ServiceWorker/#navigator-service-worker-register

If webkit uses addRegistration currently that's due to a bug, and will be fixed by https://github.com/WebKit/WebKit/pull/27880

**Review by lukewarlow — CHANGES_REQUESTED** [review]

**Review by lukewarlow — APPROVED** [review]

File: LayoutTests/imported/w3c/web-platform-tests/trusted-types/worker-constructor.https.html [hunk 6]
```diff
@@ -64,7 +64,16 @@
 // Tests with default policy.
 promise_test(t => {
   trustedTypes.createPolicy("default", {
-      createScriptURL: s => s.replace("potato", "https") });
+    createScriptURL: (s, _, sink) => {
+      if (s === "Worker") {
+        assert_equals(sink, 'Worker create');
```
Comment (lukewarlow) [inline LayoutTests/imported/w3c/web-platform-tests/trusted-types/worker-constructor.https.html]:
```suggestion
        assert_equals(sink, 'Worker constructor');
```

File: LayoutTests/imported/w3c/web-platform-tests/trusted-types/worker-constructor.https.html [hunk 7]
```diff
@@ -64,7 +64,16 @@
...
+      } else if (s === "SharedWorker") {
+        assert_equals(sink, 'SharedWorker create');
```
Comment (lukewarlow) [inline LayoutTests/imported/w3c/web-platform-tests/trusted-types/worker-constructor.https.html]:
```suggestion
        assert_equals(sink, 'SharedWorker constructor');
```

**Review by lukewarlow — CHANGES_REQUESTED** [review]

**Review by lukewarlow — APPROVED** [review]

**Review by mrego — APPROVED** [review]

File: LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-DOMWindowTimers-setTimeout-setInterval.html [hunk 8]
```diff
@@ -54,7 +54,14 @@
     });
   }, "`window.setInterval(null)` throws.");
 
-  let policy = window.trustedTypes.createPolicy("default", { createScript: x => "0" });
```
Comment (ghost) [inline LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-DOMWindowTimers-setTimeout-setInterval.html:57]: @ziransun, @lukewarlow, @mrego: FYI those two assertions are never checked because "timeoutTestString"/"intervalTestString" is never passed to an injection sink.

Comment (ghost) [reply LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-DOMWindowTimers-setTimeout-setInterval.html:57]: Adding a corresponding test in https://phabricator.services.mozilla.com/D232984.

File: LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-duplicate-names-without-enforcement.html [hunk 9]
```diff
@@ -6,6 +6,10 @@
 <body>
 <script>
 const policy = { createHTML: a => a };
+const policy_default = { createHTML: (a, _, sink) => {
+  assert_equals(sink, 'Element innerHTML');
+  return a;
+}};
```
Comment (fred-wang) [inline LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-duplicate-names-without-enforcement.html:12]: Similarly to what @mbrodesser-Igalia  commented above, this one will never be executed in that test. Fortunately, this sink is already covered by your change from trusted-types-createHTMLDocument.html above.

Anyway, I think we should revert the change to trusted-types/trusted-types-duplicate-names-without-enforcement.html here.

cc @lukewarlow
